### ISSUE TITLE
graceful scale down

### ISSUE THREAD

**Issue by rashtao** [Unitech/pm2#1346]
When scaling down the children are killed and not gracefully shutdown. 

Eg. In my app code I have: 

``` javascript
    process.on('message', function (msg) {
        if (msg == 'shutdown') {
            // log something to file
        }
    });
```

I launch it with pm2:

``` bash
$ pm2 start app.js -i 4
[PM2] Process app.js launched
┌──────────┬────┬─────────┬───────┬────────┬─────────┬────────┬─────────────┬──────────┐
│ App name │ id │ mode    │ pid   │ status │ restart │ uptime │ memory      │ watching │
├──────────┼────┼─────────┼───────┼────────┼─────────┼────────┼─────────────┼──────────┤
│ app      │ 93 │ cluster │ 20165 │ online │ 0       │ 110s   │ 25.973 MB   │ disabled │
│ app      │ 94 │ cluster │ 20167 │ online │ 0       │ 110s   │ 25.969 MB   │ disabled │
│ app      │ 95 │ cluster │ 20207 │ online │ 0       │ 105s   │ 25.938 MB   │ disabled │
│ app      │ 96 │ cluster │ 20212 │ online │ 0       │ 105s   │ 25.938 MB   │ disabled │
└──────────┴────┴─────────┴───────┴────────┴─────────┴────────┴─────────────┴──────────┘
```

and I give the command:

``` bash
$ pm2 scale app 3
```

one child is terminated, but the message event is not fired. So I guess that the child is brutally killed.

The message events are correctly fired when I do:

``` bash
$ pm2 gracefulReload app
[PM2] Reloading process by name app
[PM2] Process app succesfully reloaded
[PM2] Process app succesfully reloaded
[PM2] All processes reloaded
```

What am I missing? 
Is it possible to "gracefullly" scale down?

**Comment by Fantus** [comment]
i do have the same issue, as i need to close some connection before the worker process gets killed. any suggestions?

**Comment by jtokoph** [comment]
I know it's been a while since this was closed, but nobody gave any alternative for gracefulStop.

After a bit of digging I found a solution and wanted to post it here, as my Google searches brought me here first.

`gracefulReload` uses `process.on('message')` to catch the reload notification. The rest of the stops and restarts send a `SIGINT` as documented here: http://pm2.keymetrics.io/docs/usage/signals-clean-restart/

Simple solution is to call your same clean shutdown code in a `SIGINT` handler as documented at the same link. If you want your process to be given more than the default 1.6 seconds to cleanly shutdown, you'll need to set the `PM2_KILL_TIMEOUT` environment variable to the number of milliseconds you want.
